### PR TITLE
Backend check changes and huawei CRUD

### PR DESCRIPTION
**Pull request by vineela1999**
This PR includes
1. CRUD  bucket implementation of huawei cloud and validation of credentials of aws and azure backends
2. Includes multipart object upload of aws,azure, huawei cloud.

Given fake access key:
Screenshot from 2021-03-14 13-38-59

Showing error as **invalid credentials**:
Screenshot from 2021-03-14 13-39-35

### CODE REVIEW

File: s3/pkg/datastore/gcp/gcp.go [hunk 1]
```diff
@@ -53,8 +53,8 @@ func (ad *GcsAdapter) BucketDelete(ctx context.Context, input *pb.Bucket) error
 
 func (ad *GcsAdapter) BucketCreate(ctx context.Context, input *pb.Bucket) error {
 	bucket := ad.session.NewBucket()
-	acl := models.ACL(input.Acl.CannedAcl)
-	err := bucket.Create(input.Name, acl)
+	//acl := models.ACL(input.Acl.CannedAcl)
```
Comment (PravinRanjan10) [inline s3/pkg/datastore/gcp/gcp.go:56]: delete this line

File: s3/pkg/datastore/azure/azure.go [hunk 2]
```diff
@@ -52,7 +53,7 @@ func (ad *AzureAdapter) BucketDelete(ctx context.Context, input *pb.Bucket) erro
 
 	_, err := containerURL.Delete(ctx, azblob.ContainerAccessConditions{})
 	if err != nil {
-		log.Error("failed to delete bucket:",err)
+		log.Error("failed to delete bucket:", err)
```
Comment (PravinRanjan10) [inline s3/pkg/datastore/azure/azure.go:56]: remove extra space near to "err"

File: s3/pkg/datastore/azure/azure.go [hunk 3]
```diff
@@ -66,7 +67,7 @@ func (ad *AzureAdapter) BucketCreate(ctx context.Context, input *pb.Bucket) erro
 	// Create the container on the service (with no metadata and no public access)
 	_, err := containerURL.Create(ctx, azblob.Metadata{}, azblob.PublicAccessNone)
 	if err != nil {
-		log.Error("failed to create bucket:",err)
+		log.Error("failed to create bucket:", err)
```
Comment (PravinRanjan10) [inline s3/pkg/datastore/azure/azure.go:70]: same here

File: s3/pkg/datastore/gcp/gcp.go (hunk 1 above)
Comment (vineela1999) [reply s3/pkg/datastore/gcp/gcp.go]: done

File: s3/pkg/datastore/azure/azure.go (hunk 2 above)
Comment (vineela1999) [reply s3/pkg/datastore/azure/azure.go]: changed

File: s3/pkg/datastore/azure/azure.go (hunk 3 above)
Comment (vineela1999) [reply s3/pkg/datastore/azure/azure.go]: changed

File: s3/pkg/datastore/gcp/gcp.go [hunk 4]
```diff
@@ -53,8 +53,7 @@ func (ad *GcsAdapter) BucketDelete(ctx context.Context, input *pb.Bucket) error
 
 func (ad *GcsAdapter) BucketCreate(ctx context.Context, input *pb.Bucket) error {
 	bucket := ad.session.NewBucket()
-	acl := models.ACL(input.Acl.CannedAcl)
-	err := bucket.Create(input.Name, acl)
+	err := bucket.Create(input.Name, models.PublicReadWrite)
```
Comment (kumarashit) [inline s3/pkg/datastore/gcp/gcp.go:56]: this change doesn't seem to be relevant to the PR

File: s3/pkg/datastore/gcp/gcp.go [hunk 5]
```diff
@@ -136,7 +135,7 @@ func (ad *GcsAdapter) Get(ctx context.Context, object *pb.Object, start int64, e
 
 func (ad *GcsAdapter) Delete(ctx context.Context, input *pb.DeleteObjectInput) error {
 	bucket := ad.session.NewBucket()
-	objectId := input.Bucket + "/" + input.Key
+	objectId := input.Key
```
Comment (kumarashit) [inline s3/pkg/datastore/gcp/gcp.go:138]: Same here

Comment (vineela1999) [reply s3/pkg/datastore/gcp/gcp.go:138]: this is not relevant to PR but its a bug fix

**Comment by PravinRanjan10** [comment]
@vineela1999 Can you Update the PR description?. Please include what all fixes are in this PR.

File: s3/pkg/datastore/gcp/gcp.go (hunk 5 above)
Comment (PravinRanjan10) [reply s3/pkg/datastore/gcp/gcp.go:138]: Plz update the PR description, mention what all fixes are included

File: s3/pkg/datastore/gcp/gcp.go (hunk 4 above)
Comment (kumarashit) [reply s3/pkg/datastore/gcp/gcp.go:56]: Why Public RW?

File: s3/pkg/datastore/huawei/huawei.go [hunk 6]
```diff
@@ -38,16 +38,31 @@ type OBSAdapter struct {
 }
 
 func (ad *OBSAdapter) BucketDelete(ctx context.Context, in *pb.Bucket) error {
+	_, err := ad.client.DeleteBucket(in.Name)
+	if err != nil {
+		log.Errorf("bucket deletion is failed [%d]\n", err)
```
Comment (kumarashit) [inline s3/pkg/datastore/huawei/huawei.go:43]: is error returned int?

File: s3/pkg/datastore/huawei/huawei.go [hunk 7]
```diff
@@ -38,16 +38,31 @@ type OBSAdapter struct {
...
 	return nil
 }
 
 func (ad *OBSAdapter) BucketCreate(ctx context.Context, input *pb.Bucket) error {
+	in := &obs.CreateBucketInput{}
+	in.Bucket = input.Name
+	in.ACL = obs.AclType(input.Acl.CannedAcl)
+	_, err := ad.client.CreateBucket(in)
+	if err != nil {
+		log.Errorf("bucket creation is failed [%d]\n", err)
```
Comment (kumarashit) [inline s3/pkg/datastore/huawei/huawei.go:56]: same here

**Review by kumarashit — CHANGES_REQUESTED** [review]

File: s3/pkg/datastore/gcp/gcp.go (hunk 5 above)
Comment (vineela1999) [reply s3/pkg/datastore/gcp/gcp.go:138]: done

File: s3/pkg/datastore/huawei/huawei.go (hunk 6 above)
Comment (vineela1999) [reply s3/pkg/datastore/huawei/huawei.go]: changed

File: s3/pkg/datastore/huawei/huawei.go (hunk 7 above)
Comment (vineela1999) [reply s3/pkg/datastore/huawei/huawei.go]: changed

File: s3/pkg/datastore/gcp/gcp.go (hunk 4 above)
Comment (vineela1999) [reply s3/pkg/datastore/gcp/gcp.go:56]: since private buckets r not allowing to upload objects

File: s3/pkg/datastore/azure/azure.go [hunk 8]
```diff
@@ -306,10 +307,11 @@ func (ad *AzureAdapter) Base64ToInt64(base64ID string) int64 {
 
 func (ad *AzureAdapter) UploadPart(ctx context.Context, stream io.Reader, multipartUpload *pb.MultipartUpload,
 	partNumber int64, upBytes int64) (*model.UploadPartResult, error) {
-	bucket := ad.backend.BucketName
+	bucket := multipartUpload.Bucket
 	log.Infof("upload part[Azure Blob], bucket:%s, objectId:%s, partNumber:%d\n", bucket, multipartUpload.ObjectId, partNumber)
 
-	blobURL := ad.containerURL.NewBlockBlobURL(multipartUpload.ObjectId)
+	containerURL, _ := ad.createBucketContainerURL(ad.backend.Access, ad.backend.Security, multipartUpload.Bucket)
```
Comment (AmitRoushan) [inline s3/pkg/datastore/azure/azure.go:313]: returned error ignore ?

File: s3/pkg/datastore/azure/azure.go [hunk 9]
```diff
@@ -328,14 +330,15 @@ func (ad *AzureAdapter) UploadPart(ctx context.Context, stream io.Reader, multip
...
-	bucket := ad.backend.BucketName
+	bucket := multipartUpload.Bucket
 	result := model.CompleteMultipartUploadResult{}
 	result.Bucket = multipartUpload.Bucket
 	result.Key = multipartUpload.Key
 	result.Location = ad.backend.Name
 	log.Infof("complete multipart upload[Azure Blob], bucket:%s, objectId:%s\n", bucket, multipartUpload.ObjectId)
 
-	blobURL := ad.containerURL.NewBlockBlobURL(multipartUpload.ObjectId)
+	containerURL, _ := ad.createBucketContainerURL(ad.backend.Access, ad.backend.Security, multipartUpload.Bucket)
```
Comment (AmitRoushan) [inline s3/pkg/datastore/azure/azure.go:340]: returned error ignore ?

File: s3/pkg/datastore/azure/azure.go [hunk 10]
```diff
@@ -377,33 +380,25 @@ func (ad *AzureAdapter) ListParts(ctx context.Context, multipartUpload *pb.ListP
 }
 
 func (ad *AzureAdapter) BackendCheck(ctx context.Context, backendDetail *pb.BackendDetailS3) error {
-
-	object := &pb.Object{
-		BucketName: backendDetail.BucketName,
-		ObjectKey:  "emptyContainer/",
+	randId := uuid.NewV4().String()
+	input := &pb.Bucket{
+		Name: "sample" + randId,
```
Comment (AmitRoushan) [inline s3/pkg/datastore/azure/azure.go:385]: Please use constants for "sample"

File: s3/pkg/datastore/azure/azure.go [hunk 11]
```diff
@@ -377,33 +380,25 @@ func (ad *AzureAdapter) ListParts(ctx context.Context, multipartUpload *pb.ListP
...
 
-	bs := []byte{0}
-	stream := bytes.NewReader(bs)
-
-	_, err := ad.Put(ctx, stream, object)
-
+	err := ad.BucketCreate(ctx, input)
 	if err != nil {
-		log.Debug("failed to put object[Azure Blob]:", err)
+		log.Error("Failed to create sample bucket :", err)
```
Comment (AmitRoushan) [inline s3/pkg/datastore/azure/azure.go:390]: nit: error logs should start with small character as per convention

File: s3/pkg/datastore/huawei/huawei.go [hunk 12]
```diff
@@ -38,16 +38,31 @@ type OBSAdapter struct {
 }
 
 func (ad *OBSAdapter) BucketDelete(ctx context.Context, in *pb.Bucket) error {
+	_, err := ad.client.DeleteBucket(in.Name)
+	if err != nil {
+		log.Errorf("bucket deletion is failed :\n", err)
```
Comment (AmitRoushan) [inline s3/pkg/datastore/huawei/huawei.go:43]: nit: please remove redundant ```\n``` from added code

**Review by AmitRoushan — CHANGES_REQUESTED** [review]

File: s3/pkg/datastore/azure/azure.go (hunk 8 above)
Comment (vineela1999) [reply s3/pkg/datastore/azure/azure.go]: changed

File: s3/pkg/datastore/azure/azure.go (hunk 9 above)
Comment (vineela1999) [reply s3/pkg/datastore/azure/azure.go]: changed

File: s3/pkg/datastore/azure/azure.go (hunk 10 above)
Comment (vineela1999) [reply s3/pkg/datastore/azure/azure.go]: updated

File: s3/pkg/datastore/azure/azure.go (hunk 11 above)
Comment (vineela1999) [reply s3/pkg/datastore/azure/azure.go]: updated

File: s3/pkg/datastore/huawei/huawei.go (hunk 12 above)
Comment (vineela1999) [reply s3/pkg/datastore/huawei/huawei.go:43]: removed

**Review by kumarashit — APPROVED** [review]
LGTM

File: s3/pkg/datastore/huawei/huawei.go (hunk 12 above)
Comment (AmitRoushan) [reply s3/pkg/datastore/huawei/huawei.go:43]: still ```\n``` in log statement

Comment (AmitRoushan) [reply s3/pkg/datastore/huawei/huawei.go:43]: still can see ```\n``` in log statement :)

**Review by AmitRoushan — APPROVED** [review]
LGTM
